### PR TITLE
Warn when using non doo file as a library

### PR DESCRIPTION
**Pull request by keyboardDrummer**
### Changes
- Warn when passing non-doo files to --library
- Lots of code was made asynchronous. These locations are performing IO, so it's good that they're async now.

### How has this been tested?
- Updated existing CLI tests

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/DafnyCore/Pipeline/Compilation.cs [hunk 1]
```diff
@@ -150,29 +153,47 @@ CompilationInput input
...
         result.Add(file);
       }
     }
 
+    if (unverifiedLibrary != null) {
+      errorReporter.Warning(MessageSource.Project, "", Project.StartingToken,
+        $"The file '{Options.GetPrintPath(unverifiedLibrary)}' was passed to --library. " +
+        $"Verification previously done for that file might not be sufficient, " +
+        $"because it might have used options incompatible with the current ones. " +
+        $"Use a .doo file to enable Dafny to check that compatible options were used");
```
Comment (robin-aws) [inline Source/DafnyCore/Pipeline/Compilation.cs:194]:
```suggestion
```suggestion
        $"The file '{Options.GetPrintPath(unverifiedLibrary)}' was passed to --library. " +
        $"Verification for that file might have used options incompatible with the current ones, or might have been skipped entirely. " +
        $"Use a .doo file to enable Dafny to check that compatible options were used.");
```
```

**Review by robin-aws — CHANGES_REQUESTED** [review]

File: Source/DafnyCore/Pipeline/Compilation.cs [hunk 2]
```diff
@@ -150,29 +153,47 @@ CompilationInput input
...
-        result.Add(DafnyFile.CreateAndValidate(errorReporter, OnDiskFileSystem.Instance, Options, stdlibDooUri, Project.StartingToken));
+        var targetSpecificFile = await DafnyFile.CreateAndValidate(errorReporter, OnDiskFileSystem.Instance, Options, stdlibDooUri, Project.StartingToken);
+        if (targetSpecificFile != null) {
+          result.Add(targetSpecificFile);
+        }
       }
 
-      result.Add(DafnyFile.CreateAndValidate(errorReporter, fileSystem, Options, DafnyMain.StandardLibrariesDooUri, Project.StartingToken));
+      var file = await DafnyFile.CreateAndValidate(errorReporter, fileSystem, Options, DafnyMain.StandardLibrariesDooUri, Project.StartingToken);
+      if (file != null) {
```
[the remark this answers is not in the document]
Comment (keyboardDrummer) [reply Source/DafnyCore/Pipeline/Compilation.cs:170]: Oof, I don't recall why this was needed. Well, fastest way to remember is to remove it again ^,^

[the remark this answers is not in the document]
Comment (keyboardDrummer) [reply Source/DafnyCore/Pipeline/Compilation.cs]: Without it, the tests fail:

```
  Failed stdlibs/StandardLibraries_Errors.dfy [52 s]
  Error Message:
   System.AggregateException : One or more errors occurred. (Command returned non-zero exit code (1): DiffCommand TestFiles/LitTests/LitTest/stdlibs/StandardLibraries_Errors.dfy.expect /home/runner/work/dafny/dafny/dafny/Source/IntegrationTests/bin/Debug/net6.0/TestFiles/LitTests/LitTest/stdlibs/Output/StandardLibraries_Errors.dfy.tmp
Output:
AssertEqualWithDiff() Failure
Diff (changing expected into actual):
 
 Dafny program verifier finished with 1 verified, 0 errors
 
 Dafny program verifier finished with 1 verified, 0 errors
 StandardLibraries_Errors.dfy(20,20): Error: module Std does not exist (position 0 in path Std.Wrappers)
 1 resolution/type errors detected in StandardLibraries_Errors.dfy
 CLI: Error: cannot load DafnyStandardLibraries-notarget.doo: --unicode-char is set locally to False, but the library was built with True
 CLI: Error: cannot load DafnyStandardLibraries.doo: --unicode-char is set locally to False, but the library was built with True
+Encountered internal compilation exception: Object reference not set to an instance of an object.
 Please use the '--check' and/or '--print' option as doo files cannot be formatted in place.
 Please use the '--check' and/or '--print' option as doo files cannot be formatted in place.
```

[the remark this answers is not in the document]
Comment (robin-aws) [reply Source/DafnyCore/Pipeline/Compilation.cs:170]: Ah that's fair, if the file is missing it should fail more loudly. I forgot that `CreateAndValidate` only returns null if the file wasn't valid.

**Review by robin-aws — DISMISSED** [review]

**Review by MikaelMayer — APPROVED** [review]

**Pull request merged**